src/backend/distributed/connection/remote_commands.c

```diff
@@ -125,46 +125,68 @@ ReportConnectionError(MultiConnection *connection, int elevel)
 void
 ReportResultError(MultiConnection *connection, PGresult *result, int elevel)
 {
-	char *sqlStateString = PQresultErrorField(result, PG_DIAG_SQLSTATE);
-	char *messagePrimary = PQresultErrorField(result, PG_DIAG_MESSAGE_PRIMARY);
-	char *messageDetail = PQresultErrorField(result, PG_DIAG_MESSAGE_DETAIL);
-	char *messageHint = PQresultErrorField(result, PG_DIAG_MESSAGE_HINT);
-	char *messageContext = PQresultErrorField(result, PG_DIAG_CONTEXT);
+	/* PGresult must be released before leaving this function. */
+	PG_TRY();
+	{
+		char *sqlStateString = PQresultErrorField(result, PG_DIAG_SQLSTATE);
+		char *messagePrimary = PQresultErrorField(result, PG_DIAG_MESSAGE_PRIMARY);
+		char *messageDetail = PQresultErrorField(result, PG_DIAG_MESSAGE_DETAIL);
+		char *messageHint = PQresultErrorField(result, PG_DIAG_MESSAGE_HINT);
+		char *messageContext = PQresultErrorField(result, PG_DIAG_CONTEXT);
 
-	char *nodeName = connection->hostname;
-	int nodePort = connection->port;
-	int sqlState = ERRCODE_INTERNAL_ERROR;
+		char *nodeName = connection->hostname;
+		int nodePort = connection->port;
+		int sqlState = ERRCODE_INTERNAL_ERROR;
 
-	if (sqlStateString != NULL)
-	{
-		sqlState = MAKE_SQLSTATE(sqlStateString[0], sqlStateString[1], sqlStateString[2],
-								 sqlStateString[3], sqlStateString[4]);
-	}
+		if (sqlStateString != NULL)
+		{
+			sqlState = MAKE_SQLSTATE(sqlStateString[0],
+									 sqlStateString[1],
+									 sqlStateString[2],
+									 sqlStateString[3],
+									 sqlStateString[4]);
+		}
 
-	/*
-	 * If the PGresult did not contain a message, the connection may provide a
-	 * suitable top level one. At worst, this is an empty string.
-	 */
-	if (messagePrimary == NULL)
-	{
-		char *lastNewlineIndex = NULL;
+		/*
+		 * If the PGresult did not contain a message, the connection may provide a
+		 * suitable top level one. At worst, this is an empty string.
+		 */
+		if (messagePrimary == NULL)
+		{
+			char *lastNewlineIndex = NULL;
 
-		messagePrimary = PQerrorMessage(connection->pgConn);
-		lastNewlineIndex = strrchr(messagePrimary, '\n');
+			messagePrimary = PQerrorMessage(connection->pgConn);
+			lastNewlineIndex = strrchr(messagePrimary, '\n');
 
-		/* trim trailing newline, if any */
-		if (lastNewlineIndex != NULL)
+			/* trim trailing newline, if any */
+			if (lastNewlineIndex != NULL)
+			{
+				*lastNewlineIndex = '\0';
+			}
+		}
+
+		ereport(elevel, (errcode(sqlState), errmsg("%s", messagePrimary),
+						 messageDetail ? errdetail("%s", messageDetail) : 0,
+						 messageHint ? errhint("%s", messageHint) : 0,
+						 messageContext ? errcontext("%s", messageContext) : 0,
+						 errcontext("while executing command on %s:%d",
+									nodeName, nodePort)));
+	}
+	PG_CATCH();
+	{
+		if (elevel == ERROR)
 		{
-			*lastNewlineIndex = '\0';
+			PQclear(result);
```

Comment:
Why do we need the check for ERROR here? Afaics we should always do this if we throw?